sql: show hidden columns in SHOW CREATE TABLE

otan opened the pull request:
Release note (sql change): Hidden columns (created using NOT VISIBLE or
implicitly created via hash sharded indexes, lack of a primary key
definition or using REGIONAL BY ROW) will now display with `NOT VISIBLE`
annotations on SHOW CREATE TABLE.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59828)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
Looks mostly good.  I'm a bit concerned about the roundtrip-ability of these new SHOW statements.  Have you tested them?  Can we automate tests for them?

> Reviewed 37 of 38 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-:-MSnhi7oDGMkMbXGbqRk:bkaemij)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @otan)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 5719 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnal4SDFIY_oO02oLS:-MSnal4SDFIY_oO02oLT:b-8dnyyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/importccl/import_stmt_test.go#L5719)):*
> ```Go
> 		importCreate[0][0] = strings.Replace(importCreate[0][0], "boroname, geom", "boroname, rowid, geom", 1)
> 		// The rowid column is implicitly created as ALTER PRIMARY KEY only comes into effect later.
> 		// As such, remove the line.
> ```

Are we removing the line?  Or replacing it with the line we want?
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnaEu59JLqjJQ-PeQh:-MSnaEu59JLqjJQ-PeQi:bjvvc0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L679)):*
> ```
> regional_by_table_no_region                     CREATE TABLE public.regional_by_table_no_region (
>                                                 i INT8 NULL,
>                                                 rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
> ```

Do we have a test somewhere to ensure that this roundtrips?
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnf9k00uBGcxP7qjBB:-MSnf9k00uBGcxP7qjBC:bvpgcaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L358)):*
> ```
>    w INT8 NULL,
>    v JSONB NULL,
>    crdb_internal_z_shard_4 INT4 NOT VISIBLE NOT NULL AS (mod(fnv32(COALESCE(CAST(z AS STRING), '':::STRING)), 4:::INT8)) STORED,
> ```

Does this roundtrip?  If so, do we have a test for it?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> Have you tested them? Can we automate tests for them?

TestShowCreateTable does this, automation would have to be a separate PR to set that kind of thing up. it's a tricky thing to setup -- probably have to automate schema creates and run the same kind of test, or hijack logic tests to do this on `SHOW CREATE`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-:-MSz9l4WAxqq-RQqvJ95:b-rkm189)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @arulajmani)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 5719 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnal4SDFIY_oO02oLS:-MSz4mAKE9ui7u6xtvUW:bbe19c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/importccl/import_stmt_test.go#L5719)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Are we removing the line?  Or replacing it with the line we want?
</blockquote></details>

Oops, relic. Fixed.
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnaEu59JLqjJQ-PeQh:-MSz9ir80MT4vEH_kFzA:b-j9twib) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L679)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Do we have a test somewhere to ensure that this roundtrips?
</blockquote></details>

See `TestShowCreateTable` which does this
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnf9k00uBGcxP7qjBB:-MSz9fxR5zvBRMQbCQLY:b-j9twib) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L358)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Does this roundtrip?  If so, do we have a test for it?
</blockquote></details>

See `TestShowCreateTable` which does this


<!-- Sent from Reviewable.io -->

otan committed:
sql: show hidden columns in SHOW CREATE TABLE

Release note (sql change): Hidden columns (created using NOT VISIBLE or
implicitly created via hash sharded indexes, lack of a primary key
definition or using REGIONAL BY ROW) will now display with `NOT VISIBLE`
annotations on SHOW CREATE TABLE.

ajstorm reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59828)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, @arulajmani, and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnaEu59JLqjJQ-PeQh:-MT0l8JB3KTTrhQPuCP2:b70g29p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L679)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

See `TestShowCreateTable` which does this
</blockquote></details>

Not to be pedantic, but does it really show that it round-trips?  Doesn't it show that we get what we expect?  I'd assume a round-tripping test would then feed back in what we get and ensure we get back what we started with.

Also, can you add a stub for the REGIONAL BY ROW test case, with an issue, just in case we lose track of it?  Thx!


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59828)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @arulajmani)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnaEu59JLqjJQ-PeQh:-MT24fJ54MYsstyLgMwI:bu86qcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L679)):*
Yes, see https://github.com/cockroachdb/cockroach/blob/3344932f85f6b31ecbb3cb02ba29d62919917586/pkg/sql/show_test.go#L314-L328.

> Also, can you add a stub for the REGIONAL BY ROW test case, with an issue, just in case we lose track of it? Thx!

This involves refactoring this code to `ccl`, so I can't really add a nice stub for it. I am actively looking to work on this though (this is a step forwards), the issue is https://github.com/cockroachdb/cockroach/issues/59362)


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-:-MT2T2xh3J5DCAC_t88h:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @arulajmani)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59828#-MSnaEu59JLqjJQ-PeQh:-MT2SxrcBma82A6L2dMR:b-xv98eb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0770ebe9b70796fe5cd6f933fd28143cdc9041b/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L679)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yes, see https://github.com/cockroachdb/cockroach/blob/3344932f85f6b31ecbb3cb02ba29d62919917586/pkg/sql/show_test.go#L314-L328.

> Also, can you add a stub for the REGIONAL BY ROW test case, with an issue, just in case we lose track of it? Thx!

This involves refactoring this code to `ccl`, so I can't really add a nice stub for it. I am actively looking to work on this though (this is a step forwards), the issue is https://github.com/cockroachdb/cockroach/issues/59362)
</blockquote></details>

Ah, didn't catch that re-insert.  Thx.  All good.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm
